### PR TITLE
Feature/csl 2388 test passive wallet

### PR DESCRIPTION
**Pull request by matt-noonan**
## Description

This adds the "wallet worker", a layer that can asynchronously receive blocks and rollbacks from the block listener and apply them to a wallet. The worker is also responsible for batching related operations together. For example, it will combine a sequence of rollbacks and block applications into a single `switchToFork` operation.

The worker is agnostic about the kind of wallet it is managing; in fact, it does not even have a concept of a wallet. Instead, the worker is configured with a `WalletActionInterp` that describes how to execute basic wallet API functions such as `applyBlocks` and `switchToFork`. This makes it easy to test the wallet worker's behavior in a pure setting, and validate it against simpler operational models (e.g. the `Stack` operations in `WalletUnitTests`)

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CSL-2388

## Type of change

- [ ~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ x] 🛠 New feature (non-breaking change which adds functionality)
- [ ~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ~] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ~] 🔨 New or improved tests for existing code
- [ ~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
- [ x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist
- [ x] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

## QA Steps
The wallet worker will be active on any node running with `--new-wallet`, or can be tested purely as in `WalletUnitTests`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -38,6 +38,7 @@ run
 *#
 .#*
 .dir-locals.el
+.projectile
```
Comment (parsonsmatt) [inline .gitignore]: this feels like something that should go in a global gitignore, but I don't really care all that much about `.gitignore` hygiene :)

File: util/Pos/Util/Chrono.hs [hunk 2]
```diff
@@ -53,6 +57,9 @@ makeWrapped ''OldestFirst
 deriving instance Semigroup (f a) => Semigroup (NewestFirst f a)
 deriving instance Semigroup (f a) => Semigroup (OldestFirst f a)
 
+deriving instance Monoid (f a) => Monoid (NewestFirst f a)
+deriving instance Monoid (f a) => Monoid (OldestFirst f a)
```
Comment (parsonsmatt) [inline util/Pos/Util/Chrono.hs]: These instances are going to break the invariant that `NewestFirst` etc want to preserve by default. You need a precondition on this that, for any `as <> bs`, every element in `bs` is older than any element in `as`.

The semigroup instance would need to do sorting on the result list, or some kind of merge sort, but then you need a way to project the chronological aspect that you're sorting on.

File: wallet-new/src/Cardano/Wallet/Kernel/Actions.hs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  , WalletActionInterp(..)
+  , forkWalletWorker
+  , walletWorker
+  , interp
+  , interpList
+  , WalletWorkerState
+  , isInitialState
+  , hasPendingFork
+  , isValidState
+  ) where
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: style nit: 4 space indent

File: wallet-new/src/Cardano/Wallet/Kernel/Actions.hs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+makeLenses ''WalletWorkerState
+
+-- A helper function for lifting a `WalletActionInterp` through a monad transformer.
+lifted :: (Monad m, MonadTrans t) => WalletActionInterp m b -> WalletActionInterp (t m) b
+lifted i = WalletActionInterp
+    { applyBlocks  = lift . applyBlocks i
+    , switchToFork = \n bs -> lift (switchToFork i n bs)
+    , emit         = lift . emit i
+    }
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: Given a suitable `hoist` function, you could write this as `hoist lift`

File: wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs [hunk 5]
```diff
@@ -36,21 +37,22 @@ import           Pos.Block.Types (Blund)
...
+    , _pwlGetAccount     :: WalletId -> AccountIndex -> m (Maybe Account)
+    , _pwlUpdateAccount  :: WalletId -> AccountIndex -> AccountUpdate -> m Account
+    , _pwlDeleteAccount  :: WalletId -> AccountIndex -> m Bool
     -- * addresses
-    , _pwlGetAddresses  :: WalletId -> m [Address]
+    , _pwlGetAddresses   :: WalletId -> m [Address]
     -- * core API
-    , _pwlApplyBlocks   :: OldestFirst NE Blund -> m ()
+    , _pwlApplyBlocks    :: OldestFirst NE Blund -> m ()
+    , _pwlRollbackBlocks :: NewestFirst NE Blund -> m ()
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs]: To keep the diffs clean, it'd be good to write these records as:

```haskell
data Rec = Rec
    { foo
        :: Int -> IO Char
    , bar
        :: IO ()
    , -- etc
    }
```

I may just do this in a separate PR, as I seem to be the only one that cares :sweat_smile:

File: wallet-new/test/unit/WalletUnitTest.hs [hunk 6]
```diff
@@ -261,6 +264,114 @@ testActiveWallet =
...
+
+testWalletWorker :: Spec
+testWalletWorker = do
+    describe "Test wallet worker state machine" $ do
+
+      it "Starts in a valid initial state with no effect on the wallet" $ do
+        let StackResult{..} = runStackWorker [] $ Stack [1..10]
+        srState `shouldSatisfy` Actions.isValidState
+        srState `shouldSatisfy` Actions.isInitialState
+        srStack `shouldBe` Stack [1..10]
```
Comment (parsonsmatt) [inline wallet-new/test/unit/WalletUnitTest.hs]: this is using mixed 2 and 4 space indentation?

File: wallet-new/test/unit/WalletUnitTest.hs [hunk 7]
```diff
@@ -261,6 +264,114 @@ testActiveWallet =
...
+
+      it "Behaves like the simple stack model, when there is no pending fork" $ do
+        let stk0 = Stack [1..100]
+            run = (`runStackWorker` stk0)
+            doesNotResultInFork = not . Actions.hasPendingFork . srState . run
+        forAll (listOf someAction `suchThat` doesNotResultInFork) $
+          \actions -> do
+            let StackResult{..} = run actions
+                expectedStack  = execState (mapM actionToStackOp actions) stk0
+            srStack `shouldBe` expectedStack
```
Comment (parsonsmatt) [inline wallet-new/test/unit/WalletUnitTest.hs]: Awesome tests :heart:

File: wallet-new/src/Cardano/Wallet/Kernel/Actions.hs (hunk 4 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: Right. I mean something [like this](https://github.com/input-output-hk/cardano-sl/blob/develop/wallet-new/src/Cardano/Wallet/Client.hs#L181-L223). `hoist` is probably not the right word for it

**Comment by matt-noonan** [comment]
I've made a mess of this one; going to close the PR and apply the changes to a new branch.
